### PR TITLE
Z: Short circuit variable length array cmp helper

### PR DESCRIPTION
**Pull request by ehsankianifar**
If the addresses of two arrays are equal, those two arrays are considered equal and we do not need to actually compare those arrays. In [this](https://github.com/eclipse/omr/pull/6439) PR, we modified the instructions for x86 and there was a discussion to do the same for other architectures. On Z, it is more complicated and depending on the conditions and expected response, several methods implement array compare.
This PR targets array cmp helper implementation when the length is variable. the arraycmpEvaluator calls arraycmpHelper when the length is variable or larger than 4k and requested response is unsigned.

### CODE REVIEW

**Comment by ehsankianifar** [comment]
Instruction selection for this path:
```

 [     0x2aa3a8be590]	                       LG      &GPR_0018, Parm[Parm  0<parm 0 L>] ?+0(GPR15)
 [     0x2aa3a8dc3d0]	                       LG      &GPR_0020, Parm[Parm  1<parm 1 L>] ?+8(GPR15)
 [     0x2aa3a8dc710]	                       L       GPR_0024, Parm[Parm  2<parm 2 I>] ?+16(GPR15)
 [     0x2aa3a8dc8e0]	                       XGR     GPR_0017,GPR_0017	
 [     0x2aa3a8dc990]	                       Label L0016:	# (Start of internal control flow)	
 [     0x2aa3a8dca50]	                       CLGR    &GPR_0018,&GPR_0020	
 [     0x2aa3a8dcb00]	                       BRC     BH(0x8), Label L0017	
 [     0x2aa3a8dcc30]	                       AHI     GPR_0024,0xffff	
 [     0x2aa3a8dccf0]	                       BRC     BM(0x4), Label L0017	
 [     0x2aa3a8dced0]	                       LR      GPR_0025,GPR_0024	
 [     0x2aa3a8dcf80]	                       SRA     GPR_0025,8
 [     0x2aa3a8dd040]	                       BRC     BH(0x8), Label L0019	
 [     0x2aa3a8dd280]	                       Label L0018:	
 [     0x2aa3a8dd340]	                       CLC     #390 0(256,&GPR_0018),#391 0(&GPR_0020)
 [     0x2aa3a8dd400]	                       BRC     MASK7(0x6), Label L0020	
 [     0x2aa3a8dd580]	                       LA      &GPR_0018,#392 256(&GPR_0018)
 [     0x2aa3a8dd6f0]	                       LA      &GPR_0020,#393 256(&GPR_0020)
 [     0x2aa3a8dd7a0]	                       BRCT    GPR_0025,Label L0018	
 [     0x2aa3a8dd860]	                       Label L0019:	
 [     0x2aa3a8fc300]	                       LARL    GPR_0026, &<LiteralPool Base Address>	 ; LoadLitPool
 [     0x2aa3a91ca20]	                       assocreg
 [     0x2aa3a91c490]	                       EX      GPR_0024,#396 0(GPR_0026)
 POST:
 {AssignAny:GPR_0017:R} {AssignAny:&GPR_0018:R} {AssignAny:&GPR_0020:R} {AssignAny:GPR_0024:R} {AssignAny:GPR_0025:R} {AssignAny:GPR_0026:R}
 [     0x2aa3a91cad0]	                       Label L0020:	
 [     0x2aa3a91cb90]	                       BRC     MASK9(0x8), Label L0017	
 [     0x2aa3a91cc50]	                       LGHI    GPR_0017,0x2	
 [     0x2aa3a91cd10]	                       BRC     MASK3(0x2), Label L0017	
 [     0x2aa3a91cdd0]	                       LGHI    GPR_0017,0x1	
 [     0x2aa3a91d550]	                       assocreg
 [     0x2aa3a91cfb0]	                       Label L0017:	
 POST:
 {AssignAny:GPR_0017:R} {AssignAny:&GPR_0018:R} {AssignAny:&GPR_0020:R} {AssignAny:GPR_0024:R} {AssignAny:GPR_0025:R} {AssignAny:GPR_0026:R}
 [     0x2aa3a91d6b0]	                       LGFR    GPR_0017,GPR_0017	
 [     0x2aa3a91dd10]	                       assocreg
 [     0x2aa3a91d760]	                       retn    	
 POST:
 {GPR2:GPR_0017:R}

```

File: compiler/z/codegen/OMRTreeEvaluator.cpp [hunk 1]
```diff
@@ -10833,17 +10834,29 @@ OMR::Z::TreeEvaluator::arraycmpHelper(TR::Node *node,
                                                                                                                      maxLenIn256 ? 5 : (isWideChar ? (isFoldedIf ? 10 : 11) : 5 + 4),
                                                                                                                      cg);
 
+      if (!isFoldedIf && needResultReg)
+         {
+         branchDeps->addPostCondition(retValReg, TR::RealRegister::AssignAny);
+         generateRRInstruction(cg, TR::InstOpCode::getXORRegOpCode(), node, retValReg, retValReg);
```
Comment (ehsankianifar) [inline compiler/z/codegen/OMRTreeEvaluator.cpp:10840]: in all code paths we need to init result register with zero

File: compiler/z/codegen/OMRTreeEvaluator.cpp [hunk 2]
```diff
@@ -10833,17 +10834,29 @@ OMR::Z::TreeEvaluator::arraycmpHelper(TR::Node *node,
...
+         }
+
+      generateS390LabelInstruction(cg, TR::InstOpCode::label, node, cFlowRegionStart);
```
Comment (ehsankianifar) [inline compiler/z/codegen/OMRTreeEvaluator.cpp:10843]: The short circuit branch would always be the start of control flow.

File: compiler/z/codegen/OMRTreeEvaluator.cpp [hunk 3]
```diff
@@ -10833,17 +10834,29 @@ OMR::Z::TreeEvaluator::arraycmpHelper(TR::Node *node,
...
+         source2Reg,
+         TR::InstOpCode::COND_BE,
+         isIfxcmpBrCondContainEqual ? compareTarget : cFlowRegionEnd);
+
       //We could do better job in future for the known length (>=4k)
       //Currently there are some unnecessary test/load instructions generated.
 
       if (maxLenIn256)
          {
-         if (!isFoldedIf && needResultReg)
```
Comment (ehsankianifar) [inline compiler/z/codegen/OMRTreeEvaluator.cpp:10841]: we init result before short circuit. this code no longer needed.

File: compiler/z/codegen/OMRTreeEvaluator.cpp [hunk 4]
```diff
@@ -10858,15 +10871,12 @@ OMR::Z::TreeEvaluator::arraycmpHelper(TR::Node *node,
             if (isFoldedIf)
                {
                if (isIfxcmpBrCondContainEqual)
-                  setStartInternalControlFlow(cg, node, generateS390BranchInstruction(cg, TR::InstOpCode::BRC, TR::InstOpCode::COND_BL, node, compareTarget),
-                                              isStartInternalControlFlowSet);
```
Comment (ehsankianifar) [inline compiler/z/codegen/OMRTreeEvaluator.cpp:10862]: control flow always start before short circuit branch. This is no longer necessary.

**Comment by ehsankianifar** [comment]
@r30shah Please review and let me know if it looks good to you.

**Comment by ehsankianifar** [comment]
Tested for JDK8 (LoZ and zOS) Axxon 64179, test level.sanity, all tests pass
JDK11 (LoZ) Build #20215, test sanity.functional sanity.system sanity.openjdk, all tests pass.

File: compiler/z/codegen/OMRTreeEvaluator.cpp [hunk 5]
```diff
@@ -10833,17 +10834,29 @@ OMR::Z::TreeEvaluator::arraycmpHelper(TR::Node *node,
...
+      cFlowRegionStart->setStartInternalControlFlow();
+
+      generateS390CompareAndBranchInstruction(
+         cg,
+         TR::InstOpCode::getCmpLogicalRegOpCode(),
+         node,
+         source1Reg,
+         source2Reg,
+         TR::InstOpCode::COND_BE,
+         isIfxcmpBrCondContainEqual ? compareTarget : cFlowRegionEnd);
```
Comment (ehsankianifar) [inline compiler/z/codegen/OMRTreeEvaluator.cpp:10853]: set needCC to false.

**Comment by ehsankianifar** [comment]
After fixing jump instruction:
Tested for JDK8 (LoZ and zOS) Axxon 64612, test level.sanity, all tests pass
JDK11 (LoZ) Build #20268, test sanity.functional sanity.system sanity.openjdk, all tests pass.

**Review by r30shah — APPROVED** [review]
LGTM

**Comment by r30shah** [comment]
Jenkins build zos,zlinux

**Comment by r30shah** [comment]
@hzongaro  Can I request you to merge this change if you are ok with it?

**Review by hzongaro — APPROVED** [review]
I think the changes look good.  Thanks!

**Comment by hzongaro** [comment]
x86-64 macOS failure is unrelated, of course, while relevant platform-specific testing passed.  Merging.

**Comment by hzongaro** [comment]
> x86-64 macOS failure is unrelated, of course, while relevant platform-specific testing passed. Merging.

Actually, may I ask you to adjust the wording of the commit message to describe the change that's being applied?  Maybe something like the following?

```
Add short circuit test of whether two array addresses are equal for
arraycmp.  This particular case handles unsigned result and variable
length or length greater than 4K.
Short circuit branching is added to the start of instruction sequence.
```

**Comment by ehsankianifar** [comment]
Thanks @hzongaro for your help. I modified the commit message as you suggested. Please let me know if it looks good now.

**Comment by hzongaro** [comment]
@ehsankianifar, thanks for updating the commit message.  I don't know whether you've seen the [Commit guidelines](https://github.com/eclipse-openj9/openj9/blob/master/CONTRIBUTING.md#commit-guidelines), but one thing they recommend is that lines in the body of the commit message shouldn't be longer than 72 characters, if possible.  May I ask you to adjust the spacing so that each line is no longer than 72 characters?

**Pull request merged**